src/nccl_ofi_rdma.c

```diff
@@ -1374,62 +1395,50 @@ static int post_eager_copy(nccl_net_ofi_rdma_req_t *req);
-				      uint64_t num_cqes, nccl_net_ofi_rdma_ep_t *ep,
+static inline int process_completions(struct fi_cq_data_entry *cq_entry, uint64_t num_cqes, nccl_net_ofi_rdma_ep_t *ep,
 				      nccl_net_ofi_ep_rail_t *rail)
 {
 	int ret = 0;
 	nccl_net_ofi_rdma_req_t *req = NULL;
 	uint64_t comp_idx = 0, comp_flags = 0;
 
-	for (comp_idx = 0; comp_idx < num_cqes; comp_idx++) {
-		void *op_ctx = cq_entry[comp_idx].op_context;
+	rdma_req_send_data_t *send_data = NULL;
+	nccl_ofi_rdma_msg_type_t *msg_type = NULL;
 
+	for (comp_idx = 0; comp_idx < num_cqes; comp_idx++) {
+		/* The context for these operations is req.
+		 * except in the FI_REMOTE_WRITE case where is NULL */
+		req = cq_entry[comp_idx].op_context;
 		comp_flags = cq_entry[comp_idx].flags;
-
-		// TODO we don't always have a req in this function.
-		// NCCL_OFI_TRACE_COMPLETIONS(req, req);
+		assert(NULL != req || (comp_flags & FI_REMOTE_WRITE));
 
 		/**
-		 * Types of completions
-		 * 0. Connect/Accept ctrl : tagged message and connect message or connect response tag type
-		 * 1. Ctrl send complete: recv communicator AND FI_SEND
-		 * 2. Ctrl recv complete: send communicator AND FI_RECV
-		 * 5. fi_write local complete: send communicator AND FI_WRITE
-		 * 6. fi_write remote complete: recv communicator AND FI_REMOTE_WRITE
-		 * 7. flush complete      : recv communicator AND FI_READ
+		 * Types of completions:
+		 * 1. SEND: connect, connect response, or control message
+		 * 2. RECV w/o immediate data: connect, connect response, or control message
+		 * 3. RECV w/ immediate data: eager message
+		 * 4. Remote-initiated write
+		 * 5. Local-initiated write
+		 * 6. READ: flush or eager copy
 		 */
+		if (comp_flags & FI_SEND) {
+			/* Send completions */
 
-		if (OFI_UNLIKELY((comp_flags & FI_TAGGED) && !IS_DATA_MSG_TYPE(cq_entry[comp_idx].tag))) {
-			/* Type 0 */
-			assert(IS_CONN_MSG_TYPE(cq_entry[comp_idx].tag) || IS_CONN_RESP_MSG_TYPE(cq_entry[comp_idx].tag));
-
-			req = op_ctx;
-			ret = inc_req_completion(req, cq_entry[comp_idx].len, 1);
-			if (OFI_UNLIKELY(ret != 0)) {
-				return ret;
-			}
-
-			if (IS_CONN_RESP_MSG_TYPE(cq_entry[comp_idx].tag) && (comp_flags & FI_RECV)) {
-				assert(req->comm->type == NCCL_NET_OFI_SEND_COMM);
-				/* Complete send communicator */
-				nccl_net_ofi_rdma_send_comm_t *s_comm =
-					(nccl_net_ofi_rdma_send_comm_t *)req->comm;
-				assert(s_comm->conn_resp_req == req);
-				ret = finish_connect(s_comm);
-			}
-		} else if (comp_flags & FI_SEND) {
-			/* The context for these operations is req. */
-			req = op_ctx;
-
-			if (req->type == NCCL_OFI_RDMA_SEND_CTRL) {
-				/* Type 1 */
+			if (req->type == NCCL_OFI_RDMA_SEND_CONN || req->type == NCCL_OFI_RDMA_SEND_CONN_RESP) {
+				/* CONN or CONN_RESP send completion */
+				ret = inc_req_completion(req, cq_entry[comp_idx].len, 1);
+				if (OFI_UNLIKELY(ret != 0)) {
+					goto exit;
+				}
+			} else if (req->type == NCCL_OFI_RDMA_SEND_CTRL) {
```

Comment:
I know this is not your doing, but the general flow around the process_cq path and inc_req_completion() needs to be unified. Either it all happens here or inside individual handler methods.

Reply:
from what I see, if we exclude the error handling and tracepoints, all cases have 1 or 2 lines of code (function calls). Handler methods are used when there is significant code, in other cases is just `inc_req_completion`. 
I wouldn't add a bunch of 1 or 2 lines functions used only once, unless you feel strongly about it. 
But I have now removed the `if (ret != 0)` statements from each branch and do this only once at the end, which makes this function shorter.

Reply:
I was suggesting the opposite... pulling out inc_req_completion() from the handlers to be here uniformly.. not adding small handler functions for the other cases. We don't have to deal with it in this PR.